src/main/java/htsjdk/tribble/gff/Gff3Writer.java

```diff
@@ -0,0 +1,103 @@
+package htsjdk.tribble.gff;
+
+import htsjdk.samtools.util.FileExtensions;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.tribble.TribbleException;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.zip.GZIPOutputStream;
+
+
+public class Gff3Writer implements Closeable {
+
+    private final PrintStream out;
+    private final static String version = "3.1.25";
+
+    public Gff3Writer(final Path path) throws IOException {
+        if (!FileExtensions.GFF3.stream().anyMatch(e -> path.toString().endsWith(e))) {
+            throw new TribbleException("File " + path + " does not have extension consistent with gff3");
+        }
+
+        final OutputStream outputStream = IOUtil.hasGzipFileExtension(path)? new GZIPOutputStream(Files.newOutputStream(path)) : Files.newOutputStream(path);
+        out = new PrintStream(outputStream);
+        //start with version directive
+        initialize();
+    }
+
+    public Gff3Writer(final PrintStream stream) {
+        out = stream;
+        initialize();
+    }
+
+    private void initialize() {
+        out.println(Gff3Codec.Gff3Directive.VERSION3_DIRECTIVE.encode(version));
+    }
+
+    public void addFeature(final Gff3Feature feature) {
+        final String lineNoAttributes = String.join("\t",
```

Comment:
It's more performant to directly `.append` each encoded value rather than convert them to a bigger string first.  That just adds layers of internal copies.  If you think this is more readable that's fine though.

Reply:
done